Test/civl/inductive-sequentialization/PingPong.bpl

```diff
@@ -21,213 +37,254 @@ function {:inline} EmptyChannel () : [int]int
+MAIN' ({:linear_in "cid"} cid: ChannelId)
 {
-  assert ping_pid == ping_id;
-  assert pong_pid == pong_id;
-  assert ping_channel == EmptyChannel();
-  assert pong_channel == EmptyChannel();
+  assert channel[cid] == ChannelPair(EmptyChannel(), EmptyChannel());
 }
 
 procedure {:IS_invariant}{:layer 2}
-INV ({:linear_in "pid"} ping_pid:int, {:linear_in "pid"} pong_pid:int)
-returns ({:pending_async "PING","PONG"} PAs:[PA]int, {:choice} choice:PA)
-modifies ping_channel, pong_channel;
+INV ({:linear_in "cid"} cid: ChannelId)
+returns ({:pending_async "PING","PONG"} PAs: [PA]int, {:choice} choice: PA)
+modifies channel;
 {
-   var {:pool "INV"} c: int;
+  var {:pool "INV"} c: int;
 
-  assert ping_pid == ping_id;
-  assert pong_pid == pong_id;
-  assert ping_channel == EmptyChannel();
-  assert pong_channel == EmptyChannel();
+  assert channel[cid] == ChannelPair(EmptyChannel(), EmptyChannel());
 
   assume
     {:add_to_pool "INV", c, c+1}
     0 < c;
   if (*) {
-    pong_channel := EmptyChannel()[c := 1];
-    ping_channel := EmptyChannel();
-    PAs := NoPAs()[PONG(c, pong_pid) := 1][PING(c, ping_pid) := 1];
-    choice := PONG(c, pong_pid);
+    channel[cid] := ChannelPair(EmptyChannel(), EmptyChannel()[c := 1]);
+    PAs := NoPAs()[PONG(c, Right(cid)) := 1][PING(c, Left(cid)) := 1];
+    choice := PONG(c, Right(cid));
   } else if (*) {
-    pong_channel := EmptyChannel()[0 := 1];
-    ping_channel := EmptyChannel();
-    PAs := NoPAs()[PONG(c, pong_pid) := 1];
-    choice := PONG(c, pong_pid);
+    channel[cid] := ChannelPair(EmptyChannel(), EmptyChannel()[0 := 1]);
+    PAs := NoPAs()[PONG(c, Right(cid)) := 1];
+    choice := PONG(c, Right(cid));
   } else if (*) {
-    ping_channel := EmptyChannel()[c := 1];
-    pong_channel := EmptyChannel();
-    PAs := NoPAs()[PONG(c+1, pong_pid) := 1][PING(c, ping_pid) := 1];
-    choice := PING(c, ping_pid);
+    channel[cid] := ChannelPair(EmptyChannel()[c := 1], EmptyChannel());
+    PAs := NoPAs()[PONG(c+1, Right(cid)) := 1][PING(c, Left(cid)) := 1];
+    choice := PING(c, Left(cid));
   } else {
-    ping_channel := EmptyChannel();
-    pong_channel := EmptyChannel();
+    channel[cid] := ChannelPair(EmptyChannel(), EmptyChannel());
     PAs := NoPAs();
   }
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 
 procedure {:IS_abstraction}{:layer 2}
-PING' (x:int, {:linear_in "pid"} pid:int)
-returns ({:pending_async "PING"} PAs:[PA]int)
-modifies ping_channel, pong_channel;
+PING' (x: int, {:linear_in "cid"} left: Permission)
+returns ({:pending_async "PING"} PAs: [PA]int)
+modifies channel;
 {
-  assert (exists {:pool "INV"} m:int :: ping_channel[m] > 0) && (forall m:int :: pong_channel[m] == 0);
-  call PAs := PING(x, pid);
+  var cid: ChannelId;
+  var left_channel: [int]int;
+  var right_channel: [int]int;
+
+  cid := ChannelId(left);
+  left_channel := left#ChannelPair(channel[cid]);
+  right_channel := right#ChannelPair(channel[cid]);
```

Comment:
This is done in multiple places. Would be nice to extract it. Unfortunately, as far is I know a function cannot return multiple values. An auxiliary atomic action would work, but it must take `channel` as input value instead of accessing the global variable in order to not break commutativity.

Also, in this context we could call the channels `ping_channel` and `pong_channel`.

Reply:
Your comment about the extracted atomic action interfering with commutativity is interesting. Now that atomic actions are allowed to call each other, users may create many auxiliary helper atomic actions which may be called only from other atomic actions but never directly from a procedure. I am wondering if we should use an analysis inside Civl to remove such actions from the pool over which commutativity checking performed.

Reply:
Indeed, extracting into an atomic action would help

```
var cid: ChannelId;
  var left_channel: [int]int;
  var right_channel: [int]int;

  cid := ChannelId(right);
  left_channel := left#ChannelPair(channel[cid]);
  right_channel := right#ChannelPair(channel[cid]);
```
The above six lines would become a single line of code. But to get the full effect, we would need the analysis I mentioned in my last comment.

I thought about renaming left_channel to ping_channel (and right_channel to pong_channel). It didn't seem to help my understanding so I am not doing it. You are welcome to make changes as you please once this PR lands.

Reply:
> The above six lines would become a single line of code.

We would still have the declarations of the three local variables, but only one assignment.

> But to get the full effect, we would need the analysis I mentioned in my last comment.

Yes, having some mechanism to exclude helper actions from commutativity checking sounds useful. I'm wondering though if a more transparent solution (e.g., where the programmer explicitly marks such actions) would be better. Otherwise code becomes hard to read, since the programmer needs to keep in mind which actions (and potential bogus mover types) get ignored.

Either way, we'll do this separately from this PR.